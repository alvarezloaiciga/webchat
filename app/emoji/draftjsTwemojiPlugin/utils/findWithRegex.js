const findWithRegex = (regex, contentBlock, callback) => {
  // Get the text from the contentBlock
  const text = contentBlock.getText();
  let matchArr;
  let start; // eslint-disable-line
  // Go through all matches in the text and return the indices to the callback
  // eslint-disable-next-line no-cond-assign
  while ((matchArr = regex.exec(text)) !== null) {
    start = matchArr.index;
    callback(start, start + matchArr[0].length);
  }
};

export default findWithRegex;
