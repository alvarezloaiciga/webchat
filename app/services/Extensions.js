// @flow

export type ExtensionEvent = {
  eventType: string,
  data: Object,
};

let _extensionWindow: Object;
let _extensionUrl: string;

export const registerExtension = (extensionUrl: string, extensionWindow: Object) => {
  _extensionUrl = extensionUrl;
  _extensionWindow = extensionWindow;
};

export const postExtensionEvent = (event: ExtensionEvent) => {
  if (_extensionWindow) {
    _extensionWindow.postMessage(event, _extensionUrl);
  }
};
