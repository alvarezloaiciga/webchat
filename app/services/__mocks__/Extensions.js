export const registerExtension = jest.fn();
export const postExtensionEvent = jest.fn();
