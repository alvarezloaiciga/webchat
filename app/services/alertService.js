import assets from 'assets';

// Load alert sound--must be done here, not inside alert function
// $FlowIssue
const canPlayMp3 = ['probably', 'maybe'].includes(new Audio().canPlayType('audio/mp3'));
const alertFile = canPlayMp3 ? assets.alertSound : assets.alertSoundWav;

// $FlowIssue
const alertSound = new Audio(alertFile);

/**
 * Plays a sound notification. Designed for demoing sound to user.
 */
export const playSound = () => {
  alertSound.play();
};
