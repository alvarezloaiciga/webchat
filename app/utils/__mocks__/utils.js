export const getBrowserName = () => 'Chrome';
export const getMajor = () => 60;
export const getDeviceType = () => 'PC';
export const getOSName = () => 'Windows';
export const getUAInfo = () => ({
  browser: {
    name: 'Chrome',
    major: 60,
  },
  os: {
    name: 'Mac OS',
    version: 10,
  },
  device: {
    model: 'iPad',
    vendor: 'Apple',
    type: 'tablet',
  },
  cpu: {
    architecture: 'amd64',
  },
  engine: {
    name: 'WebKit',
    version: 537.36,
  },
  navigator:
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.81 Safari/537.36',
});
export const compatibilityMode = () => false;
export const isIE9 = () => false;
export const isIE10 = () => false;
export const nonCompatibleBrowser = () => false;
export const supportsFlexbox = () => true;
export const supportsSVG = () => true;
export const displayError = (error: string) => error;
export const isIEorSafari = () => false;
