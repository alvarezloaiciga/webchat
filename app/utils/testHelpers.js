// @flow
import type {Message, IntlObject, IntlMessage} from 'types';

export const getMockMessage = (
  // eslint-disable-line import/prefer-default-export
  id?: number = 0,
  overrides?: Message | {} = {},
) => {
  const message = {
    authorType: id % 2 === 0 ? 'Customer' : 'Agent',
    text: `text--${id}`,
    id: `id--${id}`,
    timestamp: id,
    type: 'Text',
  };

  return Object.assign(message, overrides);
};

/**
 * Stub out formatMessage so that it will return the defaultMessage with basic variable replacement
 */
const formatTestMessage = (msg: IntlMessage, values?: {[key: string]: string}) => {
  if (!values) {
    return msg.defaultMessage;
  }

  let returnValue = msg.defaultMessage;
  Object.keys(values).forEach(key => {
    if (values) {
      const value = values[key];
      returnValue = returnValue.replace(`{${key}}`, value);
    }
  });

  return returnValue;
};

export const TestIntlObject: IntlObject = {
  formatMessage: formatTestMessage,
  formatRelative: value => value.toString(),
  formatTime: timestamp => timestamp.toString(),
  formatDate: timestamp => timestamp.toString(),
};
