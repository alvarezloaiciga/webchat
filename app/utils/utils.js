// @flow
declare var Modernizr: Object;
import {UAParser} from 'ua-parser-js';
import type {BrowserNames, DeviceTypes, OSNames} from 'types';

const parser = new UAParser();

export const getBrowserName = (): BrowserNames => parser.getResult().browser.name;

export const getMajor = (): number => parseInt(parser.getResult().browser.major, 10);

export const getDeviceType = (): DeviceTypes => parser.getResult().device.type;

export const getOSName = (): OSNames => parser.getResult().os.name;

export const getUAInfo = () => parser.getResult();

export const compatibilityMode = () => {
  const compatList = [
    {
      name: 'IE',
      major: 10,
    },
  ];
  const {name, major} = parser.getResult().browser;
  return !!compatList.find(i => i.name === name && parseInt(major, 10) <= i.major);
};

export const isIE9 = () => getBrowserName() === 'IE' && getMajor() <= 9;
export const isIE10 = () => getBrowserName() === 'IE' && getMajor() === 10;

export const nonCompatibleBrowser = () => getBrowserName() === 'IE' && getMajor() < 9;
// Modernizr says IE10 doesn't support flexbox.
// It kind of does, at least for what we need it for... so go ahead and ignore Modernizr in that case
export const supportsFlexbox = () => isIE10() || Modernizr.flexbox;
