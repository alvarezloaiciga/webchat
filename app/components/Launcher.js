// @flow
import React, {Component} from 'react';
import {injectIntl} from 'react-intl';
import {registerIntlObject} from 'core-ui/services/i18nService';
import quiqOptions from 'Common/QuiqOptions';
import ChatContainer from './ChatContainer';
import './styles/Launcher.scss';
import QuiqChatClient from 'quiq-chat';
import * as chatActions from 'actions/chatActions';
import {
  inStandaloneMode,
  isMobile,
  isLastMessageFromAgent,
  convertToExtensionMessages,
} from 'Common/Utils';
import {ChatInitializedState, eventTypes, displayModes} from 'Common/Constants';
import {connect} from 'react-redux';
import {compose} from 'redux';
import {getMetadataForSentry} from 'utils/errorUtils';
import type {
  IntlObject,
  ChatState,
  Message,
  ChatInitializedStateType,
  ChatMetadata,
  ChatConfiguration,
} from 'types';
import {tellClient} from 'services/Postmaster';
import {playSound} from 'services/alertService';
import {postExtensionEvent} from 'services/Extensions';

type LauncherState = {
  agentsAvailable?: boolean, // Undefined means we're still looking it up
};

export type LauncherProps = {
  intl: IntlObject,
  chatContainerHidden: boolean,
  chatLauncherHidden: boolean,
  initializedState: ChatInitializedStateType,
  transcript: Array<Message>,
  muteSounds: boolean,
  messageFieldFocused: boolean,
  configuration: ChatConfiguration,

  setChatContainerHidden: (chatContainerHidden: boolean) => void,
  setChatLauncherHidden: (chatLauncherHidden: boolean) => void,
  setAgentsAvailable: (agentsAvailable: boolean) => void,
  setChatInitialized: (initialized: ChatInitializedStateType) => void,
  setWelcomeFormRegistered: () => void,
  setAgentTyping: (typing: boolean) => void,
  updateTranscript: (transcript: Array<Message>) => void,
  updatePlatformEvents: (event: Event) => void,
  newWebchatSession: () => void,
  setChatConfiguration: (configuration: ChatMetadata) => void,
  removeMessage: (messageId: string) => void,
  setIsAgentAssigned: (isAgentAssigned: boolean) => void,
};

export class Launcher extends Component<LauncherProps, LauncherState> {
  props: LauncherProps;
  state: LauncherState = {};
  updateLauncherVisibilityInterval: number;
  typingTimeout: ?number;
  autoPopTimeout: ?number;

  componentDidMount() {
    registerIntlObject(this.props.intl);
    this.registerClientCallbacks();
    this.updateLauncherVisibilityInterval = setInterval(
      this.updateLauncherState,
      quiqOptions.agentsAvailableTimer,
    );

    if (!this.props.chatLauncherHidden) {
      clearTimeout(this.autoPopTimeout);
    }

    this.init();

    // TODO: Remove this when WS are used for join/leave
    // If we're in undocked-only mode, and this is a standalone window, we need to fire a leave event whenever window is closed
    if (inStandaloneMode() && quiqOptions.displayMode === displayModes.UNDOCKED) {
      window.addEventListener('unload', () => {
        QuiqChatClient.leaveChat(true);
      });
    }
  }

  componentWillReceiveProps(nextProps: LauncherProps) {
    // Respond to change in chat visibility
    if (nextProps.chatContainerHidden !== this.props.chatContainerHidden)
      this.handleChatVisibilityChange(nextProps.chatContainerHidden);
  }

  componentWillUnmount() {
    clearInterval(this.updateLauncherVisibilityInterval);
    clearTimeout(this.typingTimeout);
    clearTimeout(this.autoPopTimeout);

    // Set to uninitialized in case app will ever be remounted and will need to run through startSession()
    this.updateInitializedState(ChatInitializedState.UNINITIALIZED);
    QuiqChatClient.stop();
  }

  updateAgentAvailability = async (): Promise<boolean> => {
    if (quiqOptions.enforceAgentAvailability && !quiqOptions.demoMode) {
      const {available} = await QuiqChatClient.checkForAgents();
      this.props.setAgentsAvailable(available);

      return available;
    }

    return true;
  };

  updateContainerHidden = (hidden: boolean) => {
    if (hidden !== this.props.chatContainerHidden) {
      this.props.setChatContainerHidden(hidden);
    }
  };

  updateInitializedState = (newState: ChatInitializedStateType) => {
    if (newState !== this.props.initializedState) {
      this.props.setChatInitialized(newState);
    }
  };

  handleNewSession = () => {
    this.props.newWebchatSession();
  };

  handleNewMessages = (transcript: Array<Message>) => {
    this.props.updateTranscript(transcript);

    if (this.props.initializedState === ChatInitializedState.INITIALIZED) {
      tellClient(eventTypes.messageArrived, {transcript});
      if (
        !this.props.muteSounds &&
        !this.props.messageFieldFocused &&
        this.props.configuration.playSoundOnNewMessage &&
        isLastMessageFromAgent(transcript)
      ) {
        playSound();
      }
    }

    postExtensionEvent({
      eventType: 'transcriptChanged',
      data: {messages: convertToExtensionMessages(transcript)},
    });
  };

  registerClientCallbacks = () => {
    QuiqChatClient.onNewMessages(this.handleNewMessages);
    QuiqChatClient.onMessageSendFailure((messageId: string) => {
      this.props.removeMessage(messageId);
    });
    QuiqChatClient.onNewEvents(this.props.updatePlatformEvents);
    QuiqChatClient.onRegistration(this.props.setWelcomeFormRegistered);
    QuiqChatClient.onAgentTyping(this.handleAgentTyping);
    QuiqChatClient.onAgentAssigned(this.props.setIsAgentAssigned);
    QuiqChatClient.onConnectionStatusChange((connected: boolean) =>
      this.updateInitializedState(
        connected ? ChatInitializedState.INITIALIZED : ChatInitializedState.DISCONNECTED,
      ),
    );
    QuiqChatClient.onError(() => this.updateInitializedState(ChatInitializedState.ERROR));
    QuiqChatClient.onErrorResolved(() =>
      this.updateInitializedState(ChatInitializedState.INITIALIZED),
    );
    QuiqChatClient.onBurn(() => this.updateInitializedState(ChatInitializedState.BURNED));
    QuiqChatClient.onNewSession(this.handleNewSession);
    QuiqChatClient.onClientInactiveTimeout(this.handleClientInactiveTimeout);
    QuiqChatClient._withSentryMetadataCallback(getMetadataForSentry);
  };

  init = async () => {
    const configuration = await QuiqChatClient.getChatConfiguration();
    this.props.setChatConfiguration(configuration);

    if (!QuiqChatClient.isUserSubscribed() && !QuiqChatClient.hasTakenMeaningfulAction()) {
      QuiqChatClient.setChatVisible(false);
    }

    // Set initial launcher visibility and agent availability states
    await this.updateLauncherState();

    // Standalone Mode
    // Never show launcher
    // Always start session, show ChatContainer
    if (inStandaloneMode()) {
      this.updateContainerHidden(false);
      await this.startSession();
      return;
    }

    // ChatContainer Visible from cookie
    // Always start session
    // Pop chat open unless we're in undocked-only mode
    if (QuiqChatClient.isChatVisible()) {
      await this.startSession();

      if (quiqOptions.displayMode !== displayModes.UNDOCKED) {
        this.updateContainerHidden(false);
      }

      return;
    }

    // User has submitted welcome form or sent message, ChatContainer not visible
    // Show launcher if transcript length > 0
    // Always start session, don't change ChatContainer
    if (QuiqChatClient.isUserSubscribed()) {
      await this.startSession();
    }

    this.handleAutoPop();
  };

  updateLauncherState = async () => {
    const sessionInProgress =
      QuiqChatClient.isChatVisible() ||
      QuiqChatClient.isUserSubscribed() ||
      QuiqChatClient.hasTakenMeaningfulAction() ||
      !this.props.chatContainerHidden ||
      inStandaloneMode() ||
      this.props.transcript.length ||
      QuiqChatClient.isRegistered();

    // We need to always call updateAgentAvailability() to keep the agentsAvailable state up to data from
    // minute to minute
    const agentsAvailable = await this.updateAgentAvailability();

    if (sessionInProgress) {
      this.props.setChatLauncherHidden(false);
      return true;
    }

    this.props.setChatLauncherHidden(!agentsAvailable);
    return agentsAvailable;
  };

  startSession = async () => {
    if (
      this.props.initializedState === ChatInitializedState.LOADING ||
      this.props.initializedState === ChatInitializedState.INITIALIZED
    )
      return;

    try {
      this.updateInitializedState(ChatInitializedState.LOADING);
      await QuiqChatClient.start();
      this.props.setIsAgentAssigned(QuiqChatClient.isAgentAssigned());
      this.updateInitializedState(ChatInitializedState.INITIALIZED);
    } catch (e) {
      this.updateInitializedState(ChatInitializedState.ERROR);
    }
  };

  handleAgentTyping = (typing: boolean) => {
    this.props.setAgentTyping(typing);

    if (this.typingTimeout) {
      clearTimeout(this.typingTimeout);
    }
    if (!typing) return;

    this.typingTimeout = setTimeout(() => {
      this.props.setAgentTyping(false);
    }, 10000);
  };

  handleAutoPop = () => {
    if (
      !isMobile() &&
      quiqOptions.displayMode !== displayModes.UNDOCKED &&
      typeof quiqOptions.autoPopTime === 'number'
    ) {
      this.autoPopTimeout = setTimeout(async () => {
        if (!await this.updateLauncherState()) return;
        await this.startSession();
        this.updateContainerHidden(false);
      }, quiqOptions.autoPopTime);
    }
  };

  handleClientInactiveTimeout = () => {
    this.updateInitializedState(ChatInitializedState.INACTIVE);
    if (!this.props.chatContainerHidden && !inStandaloneMode()) {
      this.updateContainerHidden(true);
    }
  };

  handleChatVisibilityChange = async (hidden: boolean) => {
    if (!hidden) {
      // If container has become visible at any point, we don't want to auto-pop
      clearTimeout(this.autoPopTimeout);

      // If container ever becomes visible, launch buttons need to be made visible
      this.props.setChatLauncherHidden(false);

      await this.startSession();
      QuiqChatClient.joinChat();
    } else {
      if (this.props.initializedState !== ChatInitializedState.INACTIVE) {
        QuiqChatClient.leaveChat();
      }
    }
  };

  render() {
    return <div className="Launcher">{<ChatContainer />}</div>;
  }
}

export default compose(
  injectIntl,
  // $FlowIssue
  connect(
    (state: ChatState) => ({
      chatContainerHidden: state.chatContainerHidden,
      chatLauncherHidden: state.chatLauncherHidden,
      initializedState: state.initializedState,
      transcript: state.transcript,
      welcomeFormRegistered: state.welcomeFormRegistered,
      muteSounds: state.muteSounds,
      messageFieldFocused: state.messageFieldFocused,
      configuration: state.configuration,
    }),
    chatActions,
  ),
)(Launcher);
