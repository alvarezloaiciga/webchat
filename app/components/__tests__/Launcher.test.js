// @flow
jest.mock('Common/Utils');
jest.mock('Common/QuiqOptions');
jest.mock('quiq-chat');
jest.mock('services/Extensions');

import quiqOptions from 'Common/QuiqOptions';
import React from 'react';
import {Launcher} from '../Launcher';
import {shallow} from 'enzyme';
import {TestIntlObject, getMockMessage, getMockConfiguration} from 'utils/testHelpers';
import type {ShallowWrapper} from 'enzyme';
import type {LauncherProps} from '../Launcher';
import QuiqChatClient from 'quiq-chat';
import {ChatInitializedState} from 'Common/Constants';
import {inStandaloneMode} from 'Common/Utils';
import {postExtensionEvent} from 'services/Extensions';

jest.useFakeTimers();

describe('Launcher component', () => {
  let wrapper: ShallowWrapper;
  let instance: any;
  let render;
  let init: () => void;
  let testProps: LauncherProps;

  let checkForAgentsResponse;
  let hasTakenMeaningfulActionResponse;
  let isChatVisibleResponse;

  const updateAgentsAvailable = (available?: boolean = true) => {
    checkForAgentsResponse = Promise.resolve({available});
    QuiqChatClient.checkForAgents = jest.fn(() => checkForAgentsResponse);
  };

  const updateHasTakenMeaningfulAction = (hasTakenAction?: boolean = true) => {
    hasTakenMeaningfulActionResponse = hasTakenAction;
    QuiqChatClient.hasTakenMeaningfulAction = jest.fn(() => hasTakenMeaningfulActionResponse);
  };

  const updateIsChatVisible = (visible?: boolean = true) => {
    isChatVisibleResponse = visible;
    QuiqChatClient.isChatVisible = jest.fn(() => isChatVisibleResponse);
  };

  beforeEach(() => {
    testProps = {
      intl: TestIntlObject,
      chatContainerHidden: false,
      chatLauncherHidden: false,
      initializedState: 'initialized',
      welcomeFormRegistered: true,
      muteSounds: false,
      transcript: [],
      setChatContainerHidden: jest.fn(),
      setChatLauncherHidden: jest.fn(),
      setChatInitialized: jest.fn(),
      setWelcomeFormRegistered: jest.fn(),
      setAgentTyping: jest.fn(),
      updateTranscript: jest.fn(),
      newWebchatSession: jest.fn(),
      setAgentsAvailable: jest.fn(),
      setAgentEndedConversation: jest.fn(),
      updatePlatformEvents: jest.fn(),
      messageFieldFocused: false,
      updateChatConfigurationFromMetadata: jest.fn(),
      markChatAsSpam: jest.fn(),
      removeMessage: jest.fn(),
      setIsAgentAssigned: jest.fn(),
      configuration: getMockConfiguration(),
      updatePersistentData: jest.fn(),
    };

    init = () => {
      updateAgentsAvailable();
      updateHasTakenMeaningfulAction();
      updateIsChatVisible();
    };

    render = async () => {
      QuiqChatClient.checkForAgents.mockReturnValue(checkForAgentsResponse);
      QuiqChatClient.isChatVisible.mockReturnValue(isChatVisibleResponse);
      QuiqChatClient.hasTakenMeaningfulAction.mockReturnValue(hasTakenMeaningfulActionResponse);
      wrapper = shallow(<Launcher {...testProps} />);
      instance = wrapper.instance();
      (instance: any).componentDidMount();

      await checkForAgentsResponse;
      await hasTakenMeaningfulActionResponse;
      wrapper.update();
    };

    init();
  });

  describe('onNewMessages', () => {
    it('forwards on new messages to the extensionSdk', async () => {
      testProps.initializedState = 'loading';
      await render();
      instance.handleNewMessages([]);
      jest.runTimersToTime(11000);
      expect(postExtensionEvent).toBeCalledWith({
        eventType: 'transcriptChanged',
        data: {messages: []},
      });
    });
  });

  describe('agentTyping', () => {
    it('stops typing after 10 seconds', async () => {
      await render();
      instance.handleAgentTyping(true);
      expect(testProps.setAgentTyping).toBeCalledWith(true);
      jest.runTimersToTime(11000);
      expect(testProps.setAgentTyping).toBeCalledWith(false);
    });
  });

  describe('new webchat session', () => {
    it('fires newWebchatSession action', async () => {
      await render();
      await instance.handleNewSession();
      expect(testProps.newWebchatSession.mock.calls.length).toBe(1);
    });
  });

  describe('handleChatVisibilityChange', () => {
    describe('when chat container is visible', () => {
      it('starts session and joins chat', async () => {
        await render();
        await instance.handleChatVisibilityChange(false);
        expect(testProps.setChatLauncherHidden).toBeCalledWith(false);
        expect(QuiqChatClient.joinChat).toBeCalled();
      });
    });

    describe('when chat container is not visible', () => {
      it('leaves chat', async () => {
        await render();
        await instance.handleChatVisibilityChange(true);
        expect(QuiqChatClient.leaveChat).toBeCalled();
      });
    });
  });

  describe('initial states', () => {
    describe('standaloneMode', () => {
      it('sets state for standalone', async () => {
        updateIsChatVisible(false);
        updateHasTakenMeaningfulAction(false);
        testProps.initializedState = 'uninitialized';
        (inStandaloneMode: any).mockReturnValue(() => true);
        await render();
        expect(wrapper).toMatchSnapshot();
        (inStandaloneMode: any).mockReset();
      });
    });

    describe('chatVisisble', () => {
      describe('when there is a transcript', () => {
        it('goes straight to the chat', async () => {
          updateIsChatVisible(true);
          updateHasTakenMeaningfulAction(false);
          testProps.initializedState = 'uninitialized';
          testProps.transcript = [getMockMessage()];
          await render();
          expect(testProps.setChatLauncherHidden).toBeCalledWith(false);
        });
      });
    });

    describe('when user has taken meaningful action', () => {
      it('starts a session', async () => {
        updateIsChatVisible(false);
        updateHasTakenMeaningfulAction(true);
        testProps.initializedState = 'uninitialized';
        await render();
        expect(QuiqChatClient.start).toBeCalled();
      });
    });
  });

  describe('when agents are available', () => {
    it('renders', async () => {
      await render();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('auto pop for chat', () => {
    beforeEach(async () => {
      updateIsChatVisible(false);
      updateHasTakenMeaningfulAction(false);
      testProps.chatContainerHidden = true;
      quiqOptions.autoPopTime = 200;
      await render();
    });

    describe('when agents are not available', () => {
      it('does not pop chat', () => {
        jest.runTimersToTime(200);
        wrapper.update();
        expect(testProps.setChatContainerHidden).not.toBeCalled();
      });
    });

    describe('when agents are available', () => {
      it("opens the chat even if the end user doesn't click on it", async () => {
        await render();
        jest.runTimersToTime(200);
        wrapper.update();
        expect(testProps.setAgentsAvailable).toBeCalledWith(true);
      });
    });

    describe('when chat opens before auto_pop_time', () => {
      it('clears the timer', () => {
        instance.componentWillReceiveProps({chatContainerHidden: false});
        jest.runTimersToTime(200);
        wrapper.update();
        expect(testProps.setChatContainerHidden).not.toBeCalled();
      });

      describe('when the chat starts open', () => {
        it('clears the timer', async () => {
          updateIsChatVisible(true);
          testProps.chatContainerHidden = false;
          quiqOptions.autoPopTime = 200;
          await render();
          expect(testProps.setChatContainerHidden).not.toBeCalled();
        });
      });
    });
  });

  describe('response to client inactivity timeout', () => {
    it('sets the initialized state to inactive', async () => {
      await render();
      expect(testProps.initializedState).toBe(ChatInitializedState.INITIALIZED);
      await instance.handleClientInactiveTimeout();
      expect(testProps.setChatInitialized).toHaveBeenCalledWith(ChatInitializedState.INACTIVE);
    });

    it('restarts the client when chat is toggled', async () => {
      testProps.initializedState = ChatInitializedState.INACTIVE;
      await render();
      await instance.handleChatVisibilityChange(true);
      expect(QuiqChatClient.start).toBeCalled();
    });
  });

  describe('after a minute', () => {
    beforeEach(async () => {
      updateHasTakenMeaningfulAction(false);
      updateIsChatVisible(false);
      testProps.chatContainerHidden = true;
      testProps.chatLauncherHidden = true;
      testProps.transcript = [];
      await render();
    });

    it('calls checkForAgents again', () => {
      expect(QuiqChatClient.checkForAgents).toBeCalled();
      jest.runTimersToTime(1000 * 60);
      expect(QuiqChatClient.checkForAgents).toBeCalled();
    });

    describe('when chat container is not hidden', () => {
      it('assumes agents available', () => {
        wrapper.setProps({chatContainerHidden: false});
        jest.runTimersToTime(1000 * 60);
        expect(testProps.setChatLauncherHidden).lastCalledWith(false);
      });
    });

    describe('when client is popped into standalone mode', () => {
      it('assumes agents available', () => {
        (inStandaloneMode: any).mockReturnValue(() => false);
        jest.runTimersToTime(1000 * 60);
        expect(testProps.setChatLauncherHidden).lastCalledWith(false);
      });
    });

    describe('when client hasTakenMeaningfulAction', () => {
      it('assumes agents available', () => {
        updateHasTakenMeaningfulAction(false);
        jest.runTimersToTime(1000 * 60);
        expect(testProps.setChatLauncherHidden).lastCalledWith(false);
      });
    });

    describe('when there is a transcript', () => {
      it('assumes agents available', () => {
        wrapper.setProps({transcript: [getMockMessage()]});
        jest.runTimersToTime(1000 * 60);
        expect(testProps.setChatLauncherHidden).lastCalledWith(false);
      });
    });

    describe('when welcomeForm is registered', () => {
      it('assumes agents available', () => {
        QuiqChatClient.isRegistered.mockReturnValueOnce(true);
        jest.runTimersToTime(1000 * 60);
        expect(testProps.setChatLauncherHidden).lastCalledWith(false);
      });
    });
  });
});
