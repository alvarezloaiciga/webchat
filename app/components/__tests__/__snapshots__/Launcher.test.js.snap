// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Launcher component initial states chatVisisble when there is a transcript goes straight to the chat 1`] = `
<div
  className="Launcher"
>
  <Connect(ChatContainer) />
  <Connect(ToggleChatButton)
    toggleChat={[Function]}
  />
</div>
`;

exports[`Launcher component initial states standaloneMode sets state for standalone 1`] = `
<div
  className="Launcher"
>
  <Connect(ChatContainer) />
</div>
`;

exports[`Launcher component when agents are available renders 1`] = `
<div
  className="Launcher"
>
  <Connect(ChatContainer) />
  <Connect(ToggleChatButton)
    toggleChat={[Function]}
  />
</div>
`;
